### ISSUE TITLE
X-Real-IP - Security Issue?

### ISSUE THREAD

**Issue by martin-css** [pusher/oauth2_proxy#321]
Is there a configuration option to control whether the `X-Real-IP` address is accepted? This is obviously useful when another server / load balancer is sending traffic to the proxy.

However, if the proxy is exposed on the Internet then a malicious user can set the `X-Real-IP` header and cause the logs output to contain a faked IP address.

I searched the configuration page and didn't see any obvious mention of disabling this functionality.

**Comment by ploxiln** [comment]
The `tornado.httpserver.HTTPServer` options `xheaders` is what you are looking for, and it defaults to False / disabled. http://www.tornadoweb.org/en/stable/httpserver.html#http-server

**Comment by martin-css** [comment]
Sorry, a bit poorly worded on my part. I wasn't referring to a generic proxy but this application (`oauth2_proxy`).

I can see the following in the logging package:

```golang
        client := req.Header.Get("X-Real-IP")
	if client == "" {
		client = req.RemoteAddr
	}
```

Which will always allow the `X-Real-IP` header to override the address logged. AFAIK, this header isn't blanked elsewhere and the application is using the standard `net/http` package so this is still an issue.

So if `oauth2_proxy` is exposed directly (i.e. not behind a reverse proxy) then an attacker can force logging of a fake IP address.

**Comment by ploxiln** [comment]
My bad, I had a few issues open, and somehow mixed up which repo this one was related to!

**Comment by JoelSpeed** [comment]
@martin-css I believe you are correct and this is a real issue. As far as I'm aware there is no way to protect against this. We should try and come up with a guard to prevent malicious users from abusing this logging as they currently can

**Comment by martin-css** [comment]
It's useful to support headers like `X-Real-IP` (or the more common `X-Forwarded-For`) but these are normally disabled by default & have to be explicitly enabled due to the security implications.

I'd suggest some sort of configuration value to control whether usage of these headers is allowed?

**Comment by steakunderscore** [comment]
Would it be a good idea to validate that what we are logging is actually an IP address. The intent is to at least mitigate an attacker being able to log any arbitrary string.

**Comment by martin-css** [comment]
> Would it be a good idea to validate that what we are logging is actually an IP address. The intent is to at least mitigate an attacker being able to log any arbitrary string.

I don't see the value is this. It might restrict the format of what an attacker can set but I can still set `-H 'X-Real-Ip: 8.8.8.8` to any malicious requests and an unsuspecting sysadmin might wonder what Google's public DNS server is doing to his proxy :smile: 

I've added a pull request with my suggestion for discussion.

**Comment by ploxiln** [comment]
I think this was addressed by #331 and can be closed

**Comment by martin-css** [comment]
Yes, all done now
